examples/examples/jsonrpsee_server_low_level_api.rs

```diff
@@ -0,0 +1,286 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! This example shows how to use the low-level server API
+//! in jsonrpsee.
+//!
+//! The particular example disconnects peers that exceeds
+//! the rate limit more than ten times and bans IP addr.
+//!
+//! NOTE:
+//!
+//! Enabling tower middleware in this example doesn't work,
+//! to do so then the low-level API in hyper must be used.
+
+use std::collections::HashSet;
+use std::convert::Infallible;
+use std::net::SocketAddr;
+use std::sync::atomic::AtomicU32;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use futures::FutureExt;
+use jsonrpsee::core::{async_trait, client::ClientT};
+use jsonrpsee::http_client::HttpClientBuilder;
+use jsonrpsee::proc_macros::rpc;
+use jsonrpsee::server::middleware::rpc::*;
+use jsonrpsee::server::{
+	http, ws, ConnectionGuard, Params, PingConfig, RandomIntegerIdProvider, ServerHandle, Settings, StopHandle,
+};
+use jsonrpsee::types::{ErrorObject, ErrorObjectOwned, Request};
+use jsonrpsee::ws_client::WsClientBuilder;
+use jsonrpsee::{rpc_params, MethodResponse, Methods};
+use tokio::sync::{Mutex as AsyncMutex, OwnedSemaphorePermit};
+
+use hyper::server::conn::AddrStream;
+use tokio::sync::mpsc;
+use tracing_subscriber::util::SubscriberInitExt;
+
+struct DummyRateLimit<S> {
+	service: S,
+	count: Arc<AsyncMutex<usize>>,
+	state: mpsc::Sender<()>,
+}
+
+#[async_trait]
+impl<'a, S> RpcServiceT<'a> for DummyRateLimit<S>
+where
+	S: Send + Sync + RpcServiceT<'a>,
+{
+	async fn call(&self, req: Request<'a>, t: TransportProtocol) -> MethodResponse {
+		let mut lock = self.count.lock().await;
+
+		if *lock >= 10 {
+			let _ = self.state.try_send(());
+			MethodResponse::error(req.id, ErrorObject::borrowed(-32000, "RPC rate limit", None))
+		} else {
+			let rp = self.service.call(req, t).await;
+			*lock = *lock + 1;
```

Comment:
Maybe worth a comment like "In reality, this number should be decremented again at some point"?